### PR TITLE
fix reauth deadlock by not calling Token() during reauth

### PR DESCRIPTION
**Pull request by jrperritt**
For #645

### CODE REVIEW

**Comment by jrperritt** [comment]
cc @jtopjian

**Comment by jtopjian** [comment]
I was pretty close to figuring this one out :)

It doesn't look like it's possible to detect the type of lock held? If it was possible, then the rlock could check if the wlock was currently in place, right?

**Comment by jtopjian** [comment]
oh shoot. I now see that an rlock can't be in place when there's an existing wlock. I was going to question this the other day but I assumed that it was possible. Lesson learned.

**Comment by jrperritt** [comment]
> I now see that an rlock can't be in place when there's an existing wlock

Correct. Grabbing the write lock will block all other operations. It was my fault for not putting `AuthenticationHeaders` in the unit test; that would've caught this.

**Comment by mlowery** [comment]
@jrperritt: Not at my desk to validate this claim but my concern with this change is that it introduces a field that isn’t guarded with a mutex in all cases. I think a `go test -race` unit test might be able to prove this. Thoughts?

**Comment by jrperritt** [comment]
> my concern with this change is that it introduces a field that isn’t guarded with a mutex in all cases

Yes, that's true, but I think in practice the worst case is that a goroutine may send an unauthorized request.

**Comment by jrperritt** [comment]
The commit I just pushed show what it would look like race-free.

**Comment by jrperritt** [comment]
I'm going to merge this to [hopefully] fix the bug, but I'll keep thinking about a better way to do concurrent reauth.

**Comment by mlowery** [comment]
@jrperritt: Haven't had a chance to grab the latest yet. Thanks for your prompt response on this!

**Comment by jrperritt** [comment]
@mlowery Of course, thank you for reporting it!

File: provider_client.go [hunk 1]
```diff
@@ -72,14 +72,29 @@ type ProviderClient struct {
...
 }
 
 // AuthenticatedHeaders returns a map of HTTP headers that are common for all
 // authenticated service requests.
-func (client *ProviderClient) AuthenticatedHeaders() map[string]string {
+func (client *ProviderClient) AuthenticatedHeaders() (m map[string]string) {
+	if client.reauthmut != nil {
+		client.reauthmut.RLock()
+		if client.reauthmut.reauthing {
+			client.reauthmut.RUnlock()
```
Comment (mlowery) [inline provider_client.go]: @jrperritt: Perhaps a slightly improved version would be:

```go
if client.reauthmut != nil {
    client.reauthmut.RLock()
    reauthing := client.reauthmut.reauthing
    client.reauthmut.RUnlock()
    if reauthing {
        return
    }
}
t := client.Token()
```

Comment (mlowery) [reply provider_client.go]: (Removes the need for two `RUnlock` calls.)

Comment (jrperritt) [reply provider_client.go]: > Perhaps a slightly improved version would be:

Maybe? That'd work too, but it looks like the same number of lines. Note that only one of the `RUnlock`s is ever called.

**Pull request merged**